qupath-extension-openslide/src/main/java/qupath/lib/images/servers/openslide/OpenSlide.java

```diff
@@ -0,0 +1,491 @@
+            associated.put(s, new AssociatedImage(s, this));
+        }
+
+        associatedImages = Collections.unmodifiableMap(associated);
+
+        // store info for hash and equals
+        canonicalFile = file.getCanonicalFile();
+        String quickhash1 = getProperties().get(PROPERTY_NAME_QUICKHASH1);
+        if (quickhash1 != null) {
+            hashCodeVal = (int) Long.parseLong(quickhash1.substring(0, 8), 16);
+        } else {
+            hashCodeVal = canonicalFile.hashCode();
+        }
+
+        // dispose on error, we are in the constructor
+        try {
+            checkError();
+        } catch (IOException e) {
+            dispose();
+            throw e;
+        }
+    }
+
+    // call with the reader lock held, or from the constructor
+    private void checkError() throws IOException {
+        String msg = OpenSlideJNA.INSTANCE.openslide_get_error(osr);
+
+        if (msg != null) {
+            throw new IOException(msg);
+        }
+    }
+
+    // takes the writer lock
+    public void dispose() {
+        Lock wl = lock.writeLock();
+        wl.lock();
+        try {
+            if (osr != 0) {
+                OpenSlideJNA.INSTANCE.openslide_close(osr);
+                osr = 0;
+            }
+        } finally {
+            wl.unlock();
+        }
+    }
+
+    public int getLevelCount() {
+        return levelCount;
+    }
+
+    // call with the reader lock held
+    private void checkDisposed() {
+        if (osr == 0) {
+            throw new OpenSlideDisposedException();
+        }
+    }
+
+    public long getLevel0Width() {
+        return levelWidths[0];
+    }
+
+    public long getLevel0Height() {
+        return levelHeights[0];
+    }
+
+    public long getLevelWidth(int level) {
+        return levelWidths[level];
+    }
+
+    public long getLevelHeight(int level) {
+        return levelHeights[level];
+    }
+
+    public void paintRegionOfLevel(Graphics2D g, int dx, int dy, int sx,
+                                   int sy, int w, int h, int level) throws IOException {
+        paintRegion(g, dx, dy, sx, sy, w, h, levelDownsamples[level]);
+    }
+
+    // takes the reader lock
+    public void paintRegionARGB(int[] dest, long x, long y, int level, int w,
+                                int h) throws IOException {
+        if ((long) w * (long) h > dest.length) {
+            throw new ArrayIndexOutOfBoundsException("Size of data ("
+                    + dest.length + ") is less than w * h");
+        }
+
+        if (w < 0 || h < 0) {
+            throw new IllegalArgumentException("w and h must be nonnegative");
+        }
+
+        Lock rl = lock.readLock();
+        rl.lock();
+        try {
+            checkDisposed();
+            OpenSlideJNA.INSTANCE.openslide_read_region(osr, dest, x, y, level, w, h);
+            checkError();
+        } finally {
+            rl.unlock();
+        }
+    }
+
+    public void paintRegion(Graphics2D g, int dx, int dy, long sx, long sy,
```

Comment:
If we're looking for a minimal wrapper around OpenSlide, I don't think we need this - and I don't think QuPath uses it, in favour of https://github.com/qupath/qupath/blob/main/qupath-extension-openslide/src/main/java/qupath/lib/images/servers/openslide/OpenslideImageServer.java#L278

Basically, it has quite a bit of Java-specific logic that potentially reduces efficiency if what you really want at the end is a `BufferedImage` (not to paint to an existing `Graphics`).